### ISSUE TITLE
Google Play Issue - Change master back on wifi state change

### ISSUE THREAD

**Issue by dmfrey** [MythTV-Clients/MythTV-Android-Frontend#55]
org.springframework.web.client.ResourceAccessException: I/O error: Connection to http://192.168.100.12:6544 refused; nested exception is org.apache.http.conn.HttpHostConnectException: Connection to http://192.168.100.12:6544 refused
at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:490)
at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:452)
at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:428)
at org.mythtv.services.api.myth.impl.MythTemplate.getHostName(MythTemplate.java:126)
at org.mythtv.service.dvr.BannerDownloadService.onHandleIntent(BannerDownloadService.java:77)
at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:65)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:137)
at android.os.HandlerThread.run(HandlerThread.java:60)
Caused by: org.apache.http.conn.HttpHostConnectException: Connection to http://192.168.100.12:6544 refused
at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:183)
at org.apache.http.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:164)
at org.apache.http.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:119)
at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:360)
at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:555)
at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:487)
at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:465)
at org.springframework.social.support.HttpComponentsClientHttpRequest.executeInternal(HttpComponentsClientHttpRequest.java:81)
at org.springframework.social.support.AbstractBufferingClientHttpRequest.executeInternal(AbstractBufferingClientHttpRequest.java:48)
at org.springframework.http.client.AbstractClientHttpRequest.execute(AbstractClientHttpRequest.java:63)
at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:475)
... 8 more
Caused by: java.net.ConnectException: failed to connect to /192.168.100.12 (port 6544): connect failed: EHOSTUNREACH (No route to host)
at libcore.io.IoBridge.connect(IoBridge.java:114)
at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:192)
at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:459)
at java.net.Socket.connect(Socket.java:842)
at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:119)
at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:144)
... 18 more
Caused by: libcore.io.ErrnoException: connect failed: EHOSTUNREACH (No route to host)
at libcore.io.Posix.connect(Native Method)
at libcore.io.BlockGuardOs.connect(BlockGuardOs.java:85)
at libcore.io.IoBridge.connectErrno(IoBridge.java:127)
at libcore.io.IoBridge.connect(IoBridge.java:112)
... 23 more

**Comment by dmfrey** [comment]
# Google Play User Message

Opening app off home network, but the app opened in a saved state where it had last been used on home network. Returned back through screens, app crashed

**Comment by dmfrey** [comment]
Plan:

Construct a new profile helper that when called, will query the selected location profile and, if dependent on wifi, check that wifi exists.  If not, fall back to default away profile.

**Comment by dmfrey** [comment]
Looking for consensus:

There was a previous undertaking to construct a series of DaoHelper classes for accessing the objects in the database.  This is complete, including loading the currently selected backend.

However, the original issue indicates that it should auto-switch on network change.  Looking back now, I believe this is a non-realistic scenario.  It should be user driven if you want to connect, because you may need to open a ssh tunnel before connecting the app to your backend.

We should, however, make sure the app doesn't crash if the network does go away.

Thoughts???

**Comment by billmeek** [comment]
I'd agree. If the user leaves the relative safety of their home WiFi network,
then they should manually select an Away profile that uses SSH. It's not 
very obvious, however, user's SchedulesDirect login and password are
passed in clear text when MythAPI::GetVideoSourceList is used.

And the detection of a lost network, be it on the tablet/phone, backend host
or no response on port 6544 is critical.

**Comment by dmfrey** [comment]
Agreed.

So I think we should do some testing related to does that app fail
gracefully (no app crashes) in all current implemented features when not
connected to the network.

For instance, you should be able to see program guide, upcoming and
recorded episodes, but you shouldn't be able to kick off a live stream or
video playback or even switch to mythmote.

On Thu, Feb 14, 2013 at 9:39 AM, billmeek notifications@github.com wrote:

> I'd agree. If the user leaves the relative safety of their home WiFi
> network,
> then they should manually select an Away profile that uses SSH. It's not
> very obvious, however, user's SchedulesDirect login and password are
> passed in clear text when MythAPI::GetVideoSourceList is used.
> 
> And the detection of a lost network, be it on the tablet/phone, backend
> host
> or no response on port 6544 is critical.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/MythTV-Clients/MythTV-Android-Frontend/issues/55#issuecomment-13552039.

## 

My Google Profile http://www.google.com/profiles/dmfrey

**Comment by billmeek** [comment]
For this type of test, I've just been stopping the backend (for the most part.)

The app appears to me to do two levels of checking:

```
Is the phone/tablet connected to the network
Does: GET <BackendIP>:6544/Myth/GetHostName return success
```

In other words, I'm suggesting that there is no test for backend host
access e.g. ping. Nothing wrong with that, I'm just not testing for it.
And I'm treating the whole avahi process separately.

So, to develop a test matrix, the generic 'rules' for the "no backend
access" case would be:

```
View saved data - Allowed
Activity (e.g. Refresh/Play/Delete/Edit...) - Fail and notify or Don't present the icon
```

It looks to me like the current app cares for the above well. If the host _becomes_
unavailable, then an error message prints. If the host _is_ unavailable, then the
icons (e.g. Refresh/Play...) don't even get displayed. Or, in the case of the Home/Away
screen, Home isn't presented.

If any of the above observations are wrong, please comment.

What I haven't tested is "Pulling the rug our from under the app". That is
what happens/is expected when the network connection fails while something
is in progress? I'll do some tests on that expecting error messages and no
"red" in LogCat.

**Comment by dmfrey** [comment]
All of the above were intentional.  I wanted to disable the ability to select any network-based functions while the network is down.  When they come back up, they reappear.

The last statement, I am not sure we can account for.  I expect the error to occur if the network goes away while a network operation is taking place.  We will have to see what the player is does when this happens as it is outside the scope of the app.

I feel we are adequately checking, say, when a live stream transcode status update is taking place on the episode  page.  All other network operations are being handled in services and should be gracefully dying when an error gets thrown.

**Comment by billmeek** [comment]
@dmfrey looks like this should be closed, do you agree?

Tested by turning WiFi off on my tablet and starting the app, got Status screen
message OK (that the backend can't be reached.)

WiFi on, started app, went to Recordings screen. Turned WiFi off and then
tapped the Reload icon. Got a Toast message that explains that the backend
can't be reached.

Tested on: mythtv-android-frontend-1.10.1-418-g85aa2ea

**Comment by dmfrey** [comment]
closes #55

i agree

On Tue, Aug 27, 2013 at 4:33 PM, billmeek notifications@github.com wrote:

> @dmfrey https://github.com/dmfrey looks like this should be closed, do
> you agree?
> 
> Tested by turning WiFi off on my tablet and starting the app, got Status
> screen
> message OK (that the backend can't be reached.)
> 
> WiFi on, started app, went to Recordings screen. Turned WiFi off and then
> tapped the Reload icon. Got a Toast message that explains that the backend
> can't be reached.
> 
> Tested on: mythtv-android-frontend-1.10.1-418-g85aa2ea
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/MythTV-Clients/MythTV-Android-Frontend/issues/55#issuecomment-23368273
> .

## 

My Google Profile http://www.google.com/profiles/dmfrey
